### PR TITLE
[SQL] Dataflow analysis and instrumentation for lateness

### PR DESCRIPTION
**Pull request by mihaibudiu**
Is this a user-visible change (yes/no): no

We will document this feature when it is more thoroughly tested and end-to-end integrated.
There is a lot of churn because I have decided to name the Java operators exactly like their Rust counterparts, there are simply too many to use different naming conventions.
The current analysis only was tested for an aggregation operation. No support has been implemented for joins yet.
The results of the analysis generate working Rust, but they are also not yet integrated with the trace_with_bounds API, which means that we don't currently run with finite memory.

### CODE REVIEW

**Comment by mihaibudiu** [comment]
@ryzhyk please review only the Rust parts

**Review by ryzhyk — APPROVED** [review]
I only read the Rust code.

File: sql-to-dbsp-compiler/lib/sqllib/src/binary.rs [hunk 1]
```diff
@@ -25,6 +26,11 @@ pub struct ByteArray {
     data: Vec<u8>,
 }
 
+num_entries_scalar! {
+    // TODO: Is this right?
```
[the remark this answers is not in the document]
Comment (mihaibudiu) [reply sql-to-dbsp-compiler/lib/sqllib/src/binary.rs:30]: The Rust tuple has a todo about NumEntries, should I implement it there too?

**Review by lalithsuresh — APPROVED** [review]
This went way over my head :)

**Pull request merged**